activerecord/lib/active_record/relation/predicate_builder/range_handler.rb

```diff
@@ -16,15 +16,18 @@ def initialize(predicate_builder)
       def call(attribute, value)
         begin_bind = predicate_builder.build_bind_attribute(attribute.name, value.begin)
         end_bind = predicate_builder.build_bind_attribute(attribute.name, value.end)
-        if value.begin.respond_to?(:infinite?) && value.begin.infinite?
-          if value.end.respond_to?(:infinite?) && value.end.infinite?
+
+        if impossible?(begin_bind) && value.begin > 0 || impossible?(end_bind) && value.end < 0
+          attribute.in([])
+        elsif infinity?(value.begin) || !begin_bind.value.boundable?
```

Comment:
TBH I'd like to postpone implementing like `impossible_database_value?` API (https://github.com/rails/rails/pull/30000#discussion_r162247130) in this PR.

The API is only useful for building `>` or `<` with boundable queries (like #31219), so range predicate handler is the only place where it will be used for now.

But the range predicate handler need to treat infinity value as open-ended regardless of whether + or - value. Even if `impossible_database_value?` is implemented, it does not make the code easy.

@matthewd @sgrif What do you think about that?

```diff
diff --git a/activerecord/lib/active_record/relation/predicate_builder/range_handler.rb b/activerecord/lib/active_record/relation/predicate_builder/range_handler.rb
index 7ea9b46ab6..4497d9e296 100644
--- a/activerecord/lib/active_record/relation/predicate_builder/range_handler.rb
+++ b/activerecord/lib/active_record/relation/predicate_builder/range_handler.rb
@@ -17,7 +17,7 @@ def call(attribute, value)
         begin_bind = predicate_builder.build_bind_attribute(attribute.name, value.begin)
         end_bind = predicate_builder.build_bind_attribute(attribute.name, value.end)
 
-        if impossible?(begin_bind) && value.begin > 0 || impossible?(end_bind) && value.end < 0
+        if impossible?(begin_bind) && begin_bind.value.impossible_database_value? == :above || impossible?(end_bind) && end_bind.value.impossible_database_value? == :below
           attribute.in([])
         elsif infinity?(value.begin) || !begin_bind.value.boundable?
           if infinity?(value.end) || !end_bind.value.boundable?
diff --git a/activerecord/lib/active_record/relation/query_attribute.rb b/activerecord/lib/active_record/relation/query_attribute.rb
index 6b96cf6745..f63846815b 100644
--- a/activerecord/lib/active_record/relation/query_attribute.rb
+++ b/activerecord/lib/active_record/relation/query_attribute.rb
@@ -29,6 +29,16 @@ def boundable?
       rescue ::RangeError
         @_boundable = false
       end
+
+      def impossible_database_value?
+        if boundable?
+          false
+        elsif value_before_type_cast > 0
+          :above
+        else
+          :below
+        end
+      end
     end
   end
 end
```